README.md
# Code Fields 🔢

A package that allows you to create code fields with customization options for Flutter. It can be useful for OTP and other verification methods.

<br>

<img width="320px" alt="Example image" src="https://raw.githubusercontent.com/federicodesia/code_fields/master/example/images/simple_usage.gif"/>

<br>

## ⚡️ Features

- Automatically focuses the next field on typing.
- Automatically focuses the previous field on deletation.
- Controller to clear or set a code.
- Option to change the length.
- Option to change the height and width.
- Default cursor support.
- Default error message.
- Form validation.
- Autofocus option.
- Auto close keyboard on finish option.
- On change callback function.
- On completed callback function.
- And more!

<br>

## 📌 Simple Usage

```dart
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: Text("CodeFields example"),
        ),
        body: Center(
          child: CodeFields(
            length: codeLength,
          ),
        ),
      ),
    );
  }
```

<br>

## 📌 Custom Usage
Options that allow for more control:

|  Properties  | Type | Description |
|--------------|-----------|-------------|
| `length` | int | Code length. |
| `controller` | CodeFieldsController | Allows you to clear or set a code. |
| `fieldWidth`| double | Width of each TextFormField.<br>The default width is 60px.|
| `fieldHeight`| double | Height of each TextFormField.<br>The default width is 60px.|
| `keyboardType`| TextInputType | Type of keyboard that shows up.<br>The default keyboard type is `TextInputType.number`.|
| `inputDecoration`| InputDecoration | Allows you to customize the TextFormFields.|
| `textStyle`| TextStyle | The text style of the TextFormFields.|
| `margin`| EdgeInsets | The margin between each TextFormField.<br>The default margin is `EdgeInsets.symmetric(horizontal: 8.0)`.|
| `validator`| FormFieldValidator<​String> | An optional method that validates an input. Returns an error string to display if the input is invalid, or null otherwise.<br>The returned value is displayed in the center of all TextFormField.|
| `onChanged`| Function(String) | The callback function that is executed when the entered code changes. Returns a string with the current code.|
| `onCompleted`| Function(String) | The callback function that is executed when the length of the entered code is complete. Returns a string with the current code.|
| `closeOnFinish`| bool | Option that closes the keyboard when the code is finished entering. The default value is `true`.|
| `autofocus`| bool | Select the first TextFormField and display the keyboard when it is built. The default value is `false`.|

<br>

## 📌 Changing the Style

<br>

<img width="320px" alt="Example image" src="https://raw.githubusercontent.com/federicodesia/code_fields/master/example/images/input_decoration.gif"/>

<br>

You must change the inputDecoration property:

```dart
  CodeFields(
    length: 4,
    inputDecoration: InputDecoration(
      filled: true,
      fillColor: Colors.grey.withOpacity(0.2),
                    
      enabledBorder: OutlineInputBorder(
        borderRadius: BorderRadius.circular(24.0),
        borderSide: BorderSide(color: Colors.transparent)
      ),

      border: OutlineInputBorder(
        borderRadius: BorderRadius.circular(24.0),
      ),
    ),
  )
```

<br>

## 📌 Using the Validator

<br>

<img width="320px" alt="Example image" src="https://raw.githubusercontent.com/federicodesia/code_fields/master/example/images/validation.gif"/>

<br>

#### 1. Create a function with the validation rules. Returns a string with the error or null for the normal state.
```dart
  String validateCode(String code){
    if(code.length < 4) return "Please complete the code";
    else{
      bool isNumeric = int.tryParse(code) != null;
      if(!isNumeric) return "Please enter a valid code";
    }
    return null;
  }
```

#### 2. Wrap the CodeFields widget with a Form and define a key.
```dart
  final formKey = GlobalKey<FormState>();
```

```dart
  Form(
    key: formKey,
    child: CodeFields(
      length: 4,
    )
  )
```

#### 3. Calls the form validation when the button is pressed.
```dart
  FlatButton(
    child: Text("Validate code"),
    onPressed: () {
      if(formKey.currentState.validate()){
        formKey.currentState.save();
      }
    },
  )
```

<br>

## 📌 Using the CodeFieldsController

```dart
  final CodeFieldsController _controller = new CodeFieldsController();
```

<br>

|  Functions  | Type | Description |
|--------------|-----------|-----------|
| `clearCode()` | void | Clear the code from all TextFields. |
| `setCode()` |  Function(int code) | Set a code in all TextFields. The length of the code must be equal to that defined in the widget's `length` property. |

<br>

<br>

## ⚙️ Installation

Add this to your package's pubspec.yaml file:
```yaml
dependencies:
  code_fields: ^0.0.3
```

Install it:
```yaml
$ flutter pub get
```

Import the package in your project:
```
import 'package:code_fields/code_fields.dart';
```


CHANGELOG.md
## [0.0.3] - May. 7, 2021

- Custom controller to clear or set a code.

## [0.0.2] - Nov. 23, 2020

- Callbacks bug fixed.
- Formatted code.

## [0.0.1] - Nov. 23, 2020

This is the initial release of the code_fields package.
- TextFormFields with customization options.
- Focus the previous TextFormField by pressing the delete button.
- Default cursor support.
- Validation and error message.

lib/src/code_fields.dart
import 'package:code_fields/src/code_fields_controller.dart';
import 'package:flutter/material.dart';

class CodeFields extends StatefulWidget {
  const CodeFields(
      {Key key,
      @required this.length,
      this.controller,
      this.fieldWidth = 60,
      this.fieldHeight = 60,
      this.keyboardType = TextInputType.number,
      this.inputDecoration = const InputDecoration(),
      this.textStyle,
      this.margin = const EdgeInsets.symmetric(horizontal: 8.0),
      this.validator,
      this.onChanged,
      this.onCompleted,
      this.closeOnFinish = true,
      this.autofocus = false})
      : assert(length >= 2, "The length of the code must be greater than two."),
        super(key: key);

  /// Code length.
  ///
  /// It is required and must be greater than two.
  final int length;

  /// Allows you to clear or set a code. Ensure you dispose of it when done.
  final CodeFieldsController controller;

  /// Width of each TextFormField.
  ///
  /// The default width is 60px.
  final double fieldWidth;

  /// Height of each TextFormField.
  ///
  /// The default width is 60px.
  final double fieldHeight;

  /// Type of keyboard that shows up.
  ///
  /// The default keyboard type is `TextInputType.number`.
  final TextInputType keyboardType;

  /// Allows you to customize the TextFormField.
  final InputDecoration inputDecoration;

  /// The text style of the TextFormFields.
  final TextStyle textStyle;

  /// The margin between each TextFormField.
  ///
  /// The default margin is `EdgeInsets.symmetric(horizontal: 8.0)`.
  final EdgeInsets margin;

  /// An optional method that validates an input. Returns an error string to display if the input is invalid, or null otherwise.
  ///
  /// The returned value is displayed in the center of all TextFormField.
  ///
  /// Alternating between error and normal status can cause the height of the TextFormField to change.
  ///
  /// To create a field whose height is fixed regardless of whether an error is not displayed set the [helperText] parameter to a space.
  final FormFieldValidator<String> validator;

  /// The callback function that is executed when the entered code changes.
  ///
  /// Returns a String with the current code.
  final Function(String) onChanged;

  /// The callback function that is executed when the length of the entered code is complete.
  ///
  /// Returns a String with the current code.
  final Function(String) onCompleted;

  /// Option that closes the keyboard when the code is finished entering.
  ///
  /// The default value is `true`.
  final bool closeOnFinish;

  /// Select the first TextFormField and display the keyboard when it is built.
  ///
  /// The default value is `false`.
  final bool autofocus;

  @override
  _CodeFieldsState createState() => _CodeFieldsState();
}

class _CodeFieldsState extends State<CodeFields> {
  List<TextEditingController> controllers;
  String whitespaceCharacter = "\uFEFF";
  CodeFieldsController _controller;

  String errorText = "";
  void setErrorText(String error) {
    setState(() => errorText = error);
  }

  @override
  void initState() {
    controllers = new List<TextEditingController>(widget.length);

    controllers[0] = new TextEditingController();
    for (int i = 1; i < widget.length; i++) {
      controllers[i] = new TextEditingController(text: whitespaceCharacter);
    }

    widget.controller != null ? _controller = widget.controller : _controller = new CodeFieldsController();

    // Set the functions of the controller.
    _controller.onClearCode(() => clearCode());
    _controller.onSetCode((int code) => setCode(code));

    super.initState();
  }

  @override
  void dispose() {
    controllers
        .forEach((TextEditingController controller) => controller.dispose());

    super.dispose();
  }

  void clearCode(){
    controllers[0].text = "";
    for (int i = 1; i < widget.length; i++) {
      controllers[i].text = whitespaceCharacter;
    }
  }
  
  void setCode(int code){
    String stringCode = code.toString();
    if(stringCode.length == widget.length){
      controllers[0].text = stringCode[0];
      for (int i = 1; i < widget.length; i++) {
        controllers[i].text = whitespaceCharacter + stringCode[i];
      }
    }else{
      print("[CodeFields] The length of the code must be ${widget.length}.");
    }                                                                                                                                                                                                                                       
  }

  String getCode() {
    String code = "";
    controllers.forEach((TextEditingController controller) {
      code += controller.text.replaceAll(whitespaceCharacter, "");
    });
    return code;
  }

  @override
  Widget build(BuildContext context) {
    final ThemeData themeData = Theme.of(context);

    return Column(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        Container(
          height: widget.fieldHeight + 8.0,
          child: ListView.builder(
              shrinkWrap: true,
              itemCount: controllers.length,
              scrollDirection: Axis.horizontal,
              itemBuilder: (BuildContext buildContext, int index) {
                return buildField(index);
              }),
        ),
        Visibility(
          visible: errorText != "",
          child: Text(
            errorText,
            textAlign: TextAlign.center,
            style: themeData.textTheme.caption
                .copyWith(color: themeData.errorColor)
                .merge(themeData.inputDecorationTheme.errorStyle),
            overflow: TextOverflow.ellipsis,
          ),
        )
      ],
    );
  }

  Widget buildField(int index) {
    return Container(
      height: widget.fieldHeight,
      width: widget.fieldWidth,
      margin: widget.margin,
      child: Center(
        child: TextFormField(
          controller: controllers[index],
          maxLength: index == 0 ? 1 : 2,
          textAlign: TextAlign.center,
          keyboardType: widget.keyboardType,
          style: widget.textStyle,
          enableInteractiveSelection: false,
          minLines: null,
          maxLines: null,
          expands: true,
          autofocus: widget.autofocus,
          textInputAction: index == controllers.length - 1
              ? TextInputAction.done
              : TextInputAction.next,
          onChanged: (String text) {
            if (text.isEmpty) {
              if (index > 0) {
                Future.delayed(Duration(microseconds: 1), () {
                  controllers[index].text = whitespaceCharacter;
                });

                if (index == 1)
                  controllers[index - 1].text = "";
                else
                  controllers[index - 1].text = whitespaceCharacter;
                FocusScope.of(context).previousFocus();
              }
            } else {
              if (index == controllers.length - 1 && text.length == 2) {
                if (widget.closeOnFinish) FocusScope.of(context).unfocus();
              } else if (index == 0 || text.length == 2)
                FocusScope.of(context).nextFocus();
            }

            String code = getCode();
            widget.onChanged(code);
            if (code.length == widget.length) widget.onCompleted(code);
          },
          validator: (String code) {
            String error = widget.validator.call(getCode());
            error != null ? setErrorText(error) : setErrorText("");
            return error;
          },
          decoration: widget.inputDecoration.copyWith(
            errorStyle: TextStyle(fontSize: double.minPositive),
            counter: Offstage(),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:code_fields/code_fields.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  final formKey = GlobalKey<FormState>();
  final int codeLength = 4;

  String validateCode(String code){
    if(code.length < codeLength) return "Please complete the code";
    else{
      bool isNumeric = int.tryParse(code) != null;
      if(!isNumeric) return "Please enter a valid code";
    }
    return null;
  }

  void onButtonPressed() {
    if(formKey.currentState.validate()){
      formKey.currentState.save();
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: Text("CodeFields example"),
        ),
        body: Padding(
          padding: const EdgeInsets.all(24.0),
          child: Form(
            key: formKey,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                CodeFields(
                  length: codeLength,
                  validator: validateCode,
                  inputDecoration: InputDecoration(
                    filled: true,
                    fillColor: Colors.grey.withOpacity(0.2),
                    
                    enabledBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(24.0),
                      borderSide: BorderSide(color: Colors.transparent)
                    ),

                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(24.0),
                    ),
                  ),
                ),
                SizedBox(height: 32.0,),
                SizedBox(
                  width: double.infinity,
                  child: FlatButton(
                    child: Text(
                      "Validate code",
                      style: TextStyle(
                        color: Colors.white,
                      ),
                    ),
                    height: 56.0,
                    color: Theme.of(context).primaryColor,
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(24.0),
                    ),
                    onPressed: onButtonPressed,
                  ),
                ),
              ],
            ),
          ),
        )
      ),
    );
  }
}